### DISCUSSION TITLE
mpremote: mip install errors on install to pybv11 with sd card

### DISCUSSION THREAD

**Discussion by Josverl** [micropython/micropython#14078] [General]
- mpremote v1.22.0
 - pybv1.1 with any recent firmware

A syntax error is thrown when an attempt is made to create a directory ending in a `\r`

 The problem only occurs when the sd card is inserted , 
`mip install` works fine with the same setup - but without the sd card - but then there is insufficient room for the rest of the processing

<details><summary>Full Error trace</summary>
<p>
``` py
$ mpremote mip install github:josverl/micropython-stubber/mip/mpy_v6.json
Install github:josverl/micropython-stubber/mip/mpy_v6.json
Installing github:josverl/micropython-stubber/mip/mpy_v6.json to /sd/lib
/createstubs.mpylib
Traceback (most recent call last):
  File "/home/jos/.local/bin/mpremote", line 8, in <module>
    sys.exit(main())
  File "/home/jos/.local/lib/python3.10/site-packages/mpremote/main.py", line 528, in main
    handler_func(state, args)
  File "/home/jos/.local/lib/python3.10/site-packages/mpremote/mip.py", line 184, in do_mip
    _install_package(
  File "/home/jos/.local/lib/python3.10/site-packages/mpremote/mip.py", line 147, in _install_package
    _install_json(transport, package, index, target, version, mpy)
  File "/home/jos/.local/lib/python3.10/site-packages/mpremote/mip.py", line 109, in _install_json
    _download_file(transport, _rewrite_url(url, version), fs_target_path)
  File "/home/jos/.local/lib/python3.10/site-packages/mpremote/mip.py", line 79, in _download_file
    _ensure_path_exists(transport, dest)
  File "/home/jos/.local/lib/python3.10/site-packages/mpremote/mip.py", line 33, in _ensure_path_exists
    transport.fs_mkdir(prefix)
  File "/home/jos/.local/lib/python3.10/site-packages/mpremote/transport_serial.py", line 427, in fs_mkdir
    self.exec("import os\nos.mkdir('%s')" % dir)
  File "/home/jos/.local/lib/python3.10/site-packages/mpremote/transport_serial.py", line 287, in exec
    raise TransportError("exception", ret, ret_err)
mpremote.transport.TransportError: ('exception', b'', b'Traceback (most recent call last):\r\n  File "<stdin>", line 2\r\nSyntaxError: invalid syntax\r\n')
```
</p>
</details> 
I did some debugging and I think this error stems from the below code in `mpremote/mip.py` line 166...
``` py 
            if args.target is None:
                state.transport.exec("import sys")
                lib_paths = (
                    state.transport.eval("'\\n'.join(p for p in sys.path if p.endswith('/lib'))") # fix: use | as separator
                    .decode()
                    .split("\n") # fix:  use | as separator
                )
                if lib_paths and lib_paths[0]:
                    args.target = lib_paths[0]  # Workaround : lib_paths[0].strip()
```
The `'\\n'` ,may be a typo or perhaps its a double escape. ( changing to `'\n'` did not change the behavior in any way

I think the actual error is in the choice to use `join and split("\n")`,
The string that is returned does **not** contain a single `\n` but uses `\r\n` as the newline separators ( as is standard in the MicroPython repl) 

With the current code the result is 
- `(return) SerialTransport.exec: b'/sd/lib\r\n/flash/lib\r\n'`  
 - `lib_path = ['/sd/lib\r', '/flash/lib']`  
Please notice that there now is a lingering `\r` after the first lib_paths 

I think the best fix for this would be to use another separator that cannot, or is unlikely to be used in a path ; such as "|", 
rather than work around the (CR)(LF) differences, but both can work.

@jimmo Happy to create a PR , but wanted to check first if there is more to this than I could see.

**Comment by peterhinch** [comment]
As far as I can tell `mpremote` only writes to the Flash filesystem. I use `rshell` to write to `\sd`. This is also a problem if you mount a device like (say) an EEprom chip, say to `\eep`.

It would be great if `mpremote` had a means of specifying the destination.

**Reply by Josverl** [reply]
I could specify a target with `mpremote mip install --target ... `, and that does work fine, there is just an algorithmic error.
Also as I run the same workflow (GH Actions) against multiple port, boards and versions, I prefer not to need to code against different storage layouts if I don't need to. 

I've run it with a patch to eliminate the CR in CI, and then it locates the SD card with no issue at all.
shell does not work well on Windows, but does indeed has some perks and behaviors that mpremote lacks.

I'll create a PR with the patch I'm using

**Comment by Josverl** [comment]
PR: https://github.com/micropython/micropython/pull/15332
